internal/semaphore/semaphore.go

```diff
@@ -0,0 +1,70 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: MIT
+
+package semaphore
+
+import (
+	"log"
+)
+
+type Semaphore interface {
+	//Acquire a slot in the semaphore
+	Acquire()
+
+	//Release a slot in the semaphore.
+	Release()
+
+	//GetLimit returns semaphore limit.
+	GetLimit() int
+
+	//GetCount returns current number of occupied slots in semaphore.
+	GetCount() int
+
+	//Done closes down the semaphore
+	Done()
+}
```

Comment:
Do we intend to reuse this interface elsewhere with a different underlying implementation? Do we need an interface here?

Reply:
> Do we intend to reuse this interface elsewhere with a different underlying implementation?

For ourselves, no. My intention for using interface only to centralize the Function for Semaphore. However, if there are customers using our semaphore, then that's fine too for having an interface here so no disadvantages when adding the interface.

Reply:
> if there are customers using our semaphore

I highly doubt that anyone would modify the underlying implementation of this for their own needs, in their own forks. Adding this interface does not make it any more "maintainable" than omitting it.